# 2019-12-18 (水曜日)

## djangorestframework-filters: `ImportError: cannot import name 'six' from 'django.utils'`

- https://pypi.org/project/djangorestframework-filters/
- https://github.com/philipn/django-rest-framework-filters

~~~python
  File "/Users/hide/.anyenv/envs/pyenv/versions/anaconda3-2019.10/lib/python3.7/site-packages/rest_framework_filters/__init__.py", line 3, in <module>
    from .filterset import FilterSet
  File "/Users/hide/.anyenv/envs/pyenv/versions/anaconda3-2019.10/lib/python3.7/site-packages/rest_framework_filters/filterset.py", line 8, in <module>
    from django.utils import six
ImportError: cannot import name 'six' from 'django.utils' (/Users/hide/.anyenv/envs/pyenv/versions/anaconda3-2019.10/lib/python3.7/site-packages/django/utils/__init__.py)
~~~

- https://stackoverflow.com/questions/59193514/importerror-cannot-import-name-six-from-django-utils
- https://docs.djangoproject.com/en/3.0/releases/3.0/#removed-private-python-2-compatibility-apis
- https://github.com/carltongibson/django-filter : Django 3.0+
- https://github.com/encode/django-rest-framework : Django 3.0

fix:

- Github レポジトリ の `master` をインストールすると問題ない模様

~~~py
$ pip install git+https://github.com/philipn/django-rest-framework-filters#egg=djangorestframework-filters -U
Collecting djangorestframework-filters
  Cloning https://github.com/philipn/django-rest-framework-filters to /private/var/folders/3n/0lk686q1129clzkw38blpwdc0000gp/T/pip-install-fglajf72/djangorestframework-filters
  Running command git clone -q https://github.com/philipn/django-rest-framework-filters /private/var/folders/3n/0lk686q1129clzkw38blpwdc0000gp/T/pip-install-fglajf72/djangorestframework-filters
Requirement already satisfied, skipping upgrade: djangorestframework in /Users/hide/.anyenv/envs/pyenv/versions/anaconda3-2019.10/lib/python3.7/site-packages (from djangorestframework-filters) (3.11.0)
Requirement already satisfied, skipping upgrade: django-filter>=2.0 in /Users/hide/.anyenv/envs/pyenv/versions/anaconda3-2019.10/lib/python3.7/site-packages (from djangorestframework-filters) (2.2.0)
Requirement already satisfied, skipping upgrade: django>=1.11 in /Users/hide/.anyenv/envs/pyenv/versions/anaconda3-2019.10/lib/python3.7/site-packages (from djangorestframework->djangorestframework-filters) (3.0)
Requirement already satisfied, skipping upgrade: sqlparse>=0.2.2 in /Users/hide/.anyenv/envs/pyenv/versions/anaconda3-2019.10/lib/python3.7/site-packages (from django>=1.11->djangorestframework->djangorestframework-filters) (0.3.0)
Requirement already satisfied, skipping upgrade: pytz in /Users/hide/.anyenv/envs/pyenv/versions/anaconda3-2019.10/lib/python3.7/site-packages (from django>=1.11->djangorestframework->djangorestframework-filters) (2019.3)
Requirement already satisfied, skipping upgrade: asgiref~=3.2 in /Users/hide/.anyenv/envs/pyenv/versions/anaconda3-2019.10/lib/python3.7/site-packages (from django>=1.11->djangorestframework->djangorestframework-filters) (3.2.3)
Building wheels for collected packages: djangorestframework-filters
  Building wheel for djangorestframework-filters (setup.py) ... done
  Created wheel for djangorestframework-filters: filename=djangorestframework_filters-1.0.0.dev0-cp37-none-any.whl size=20219 sha256=16be1d732f5e97c58b0e20126072ece21d2ca32b5786fb8d61a19c514baaf00f
  Stored in directory: /private/var/folders/3n/0lk686q1129clzkw38blpwdc0000gp/T/pip-ephem-wheel-cache-4l8au8gl/wheels/f0/b4/bb/3c390b5db3ab537b6c7d00225970fcfc26e243db850139495a
Successfully built djangorestframework-filters
Installing collected packages: djangorestframework-filters
  Found existing installation: djangorestframework-filters 1.0.0.dev0
    Uninstalling djangorestframework-filters-1.0.0.dev0:
      Successfully uninstalled djangorestframework-filters-1.0.0.dev0
Successfully installed djangorestframework-filters-1.0.0.dev0
~~~

## django-robots

- https://github.com/jazzband/django-robots

~~~~python
 File "/Users/hide/.anyenv/envs/pyenv/versions/anaconda3-2019.10/lib/python3.7/site-packages/robots/models.py", line 3, in <module>
    from django.utils.encoding import python_2_unicode_compatible
ImportError: cannot import name 'python_2_unicode_compatible' from 'django.utils.encoding' (/Users/hide/.anyenv/envs/pyenv/versions/anaconda3-2019.10/lib/python3.7/site-packages/django/utils/encoding.py)
~~~~

レポジトリのバージョン は問題ない模様:

~~~~bash
$ pip install git+https://github.com/jazzband/django-robots.git#egg=django-robots -U
Collecting django-robots
  Cloning https://github.com/jazzband/django-robots.git to /private/var/folders/3n/0lk686q1129clzkw38blpwdc0000gp/T/pip-install-p0gmqodq/django-robots
  Running command git clone -q https://github.com/jazzband/django-robots.git /private/var/folders/3n/0lk686q1129clzkw38blpwdc0000gp/T/pip-install-p0gmqodq/django-robots
Building wheels for collected packages: django-robots
  Building wheel for django-robots (setup.py) ... done
  Created wheel for django-robots: filename=django_robots-3.1.0-py2.py3-none-any.whl size=62939 sha256=37488f2484178fdc64e0c9702ec2cecc711faaaa6f47b96a37c11e2fa9e9d3f1
  Stored in directory: /private/var/folders/3n/0lk686q1129clzkw38blpwdc0000gp/T/pip-ephem-wheel-cache-1r0cb44s/wheels/5c/c4/f3/fa5da4954229f3bdcd25d90c754e4c33bc4de24a88133ab70e
Successfully built django-robots
Installing collected packages: django-robots
  Found existing installation: django-robots 3.1.0
    Uninstalling django-robots-3.1.0:
      Successfully uninstalled django-robots-3.1.0
Successfully installed django-robots-3.1.0
~~~


## django-markdownx

- https://github.com/neutronX/django-markdownx


~~~py
  File "/Users/hide/.anyenv/envs/pyenv/versions/anaconda3-2019.10/lib/python3.7/site-packages/markdownx/urls.py", line 9, in <module>
    from .views import (
  File "/Users/hide/.anyenv/envs/pyenv/versions/anaconda3-2019.10/lib/python3.7/site-packages/markdownx/views.py", line 5, in <module>
    from .forms import ImageForm
  File "/Users/hide/.anyenv/envs/pyenv/versions/anaconda3-2019.10/lib/python3.7/site-packages/markdownx/forms.py", line 8, in <module>
    from django.utils.six import BytesIO
ModuleNotFoundError: No module named 'django.utils.six'
~~~

以下に変更すれば大丈夫:

~~~py
from six import BytesIO
~~~

## VAgrant: Vagrantfileでの環境変数

- [Vagrantfile をコミットするために](https://qiita.com/hakomikan/items/860bb960699889849933)


- `ENV['CLIENT_ID']` のような形で環境変数にアクセス
- ヒアドキュメント: `#{ENV['CLIENT_ID']} `

~~~ruby
  config.vm.synced_folder "#{ENV['HOME']}/Documents/Dropbox/Projects", "/home/vagrant/projects"
~~~


## Boto3: IAM: ポリシー

この例だとユーザーが所属しているグループのポリシーは取得できない:

- [amazon web services - How to get the list of policies assigned to a user using boto3 for an aws profile? - Stack Overflow](https://stackoverflow.com/questions/35327051/how-to-get-the-list-of-policies-assigned-to-a-user-using-boto3-for-an-aws-profil)
